### PR TITLE
Add jsx-ban-props rule

### LINKED ISSUE #9: New Rule: jsx-ban-props
Ban a blacklist of props, such as:
- `dangerouslySetInnerHTML`

### PR DESCRIPTION
**Pull request by no23reason**
Fixes #9

### CODE REVIEW

**Comment by adidahiya** [comment]
@no23reason can you please enable CircleCI for your fork? Go to https://circleci.com/add-projects

**Comment by no23reason** [comment]
Sorry, I didn't realise I had to enable the CI. There are some issues related to other rule, so they need to be fixed first. I created #59 for that, which needs to land first to make the CI build pass.

File: src/rules/jsxBanPropsRule.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    public static FAILURE_STRING_FACTORY = (propName: string, explanation?: string) => {
+        return `Use of the prop '${propName}' is not allowed.${explanation ? " " + explanation : ""}`;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        const options = this.getOptions();
+        const walker = new JsxBanPropsWalker(sourceFile, options);
+        const propsToBan = options.ruleArguments;
+        if (propsToBan !== undefined) {
+            propsToBan.forEach(walker.addBannedProp);
```
Comment (adidahiya) [inline src/rules/jsxBanPropsRule.ts]: can you please do this kind of options handling in the walker constructor? that would make this implementation more consistent in style with other tslint core rules.

File: src/rules/jsxBanPropsRule.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        return this.applyWithWalker(walker);
+    }
+}
+
+class JsxBanPropsWalker extends Lint.RuleWalker {
+    private isInJsxAttribute = false;
+    private bannedProps: {[name: string]: string} = {};
```
Comment (adidahiya) [inline src/rules/jsxBanPropsRule.ts]: we can use an ES6 `Map<string, string>` for this. see https://github.com/palantir/tslint/pull/1984 for reference

File: src/rules/jsxBanPropsRule.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    public addBannedProp = (bannedProp: string[]) => {
+        this.bannedProps[bannedProp[0]] = bannedProp.length > 1 ? bannedProp[1] : "";
+    }
+
+    protected visitNode(node: ts.Node) {
+        if (node.kind === ts.SyntaxKind.JsxAttribute) {
```
Comment (adidahiya) [inline src/rules/jsxBanPropsRule.ts]: can you refactor this logic to use the syntax-specific visitor methods? you have [`visitJsxAttribute`](https://github.com/palantir/tslint/blob/25de3f9499b05ee7f8dbc2aaf7bf16aa9e4eec3d/src/language/walker/syntaxWalker.ts#L189) available in SyntaxWalker.

File: src/rules/jsxBanPropsRule.ts (hunk 1 above)
Comment (no23reason) [inline src/rules/jsxBanPropsRule.ts]: I wanted to, but I took inspiration in [banRule.ts](https://github.com/palantir/tslint/blob/master/src/rules/banRule.ts#L65). I guess it should be refactored as well? :)

Anyway, it is now in the constructor.

File: src/rules/jsxBanPropsRule.ts (hunk 2 above)
Comment (no23reason) [inline src/rules/jsxBanPropsRule.ts]: Yes, thank you very much for the tip, I changed it.

File: src/rules/jsxBanPropsRule.ts (hunk 3 above)
Comment (no23reason) [inline src/rules/jsxBanPropsRule.ts]: Changed, thank you for the suggestion!

**Comment by no23reason** [comment]
@adidahiya thank you for your great feedback, all of your suggestions should now be applied.

**Pull request merged**